### PR TITLE
Fix ABR stats + audit report on BOLA implementation

### PR DESCRIPTION
**Pull request by tchakabam**
concerns:

- https://www.pivotaltracker.com/n/projects/1597593/stories/130716747
- https://www.pivotaltracker.com/n/projects/1597593/stories/130624513

* Fixes ABR metrics feedback with request duration compensation for P2P data delivered to media-engine via loader
* Adds some info about BOLA ABR & how it impacts in case of live-streaming + what params we can tweak

### CODE REVIEW

File: lib/FragmentLoaderClassProvider.js [hunk 1]
```diff
@@ -12,6 +12,33 @@ function FragmentLoaderClassProvider(wrapper) {
     const FRAGMENT_LOADER_ERROR_NULL_REQUEST = 2;
     const FRAGMENT_LOADER_MESSAGE_NULL_REQUEST = 'request is null';
 
+    function getTotalBytesReceivedFromStats(stats) {
+        // NOTE: in principle this function should
+        //       be simpler (bytesReceived = stats.cdnDownloaded + stats.p2pDownloaded)
+        //       but for some reason this is the way its done in all wrappers (legacy?)
+        // TODO: we could make sure that peer-agent always simply sends a valid number in stats?
```
Comment (AxelDelmas) [inline lib/FragmentLoaderClassProvider.js]: This is probably legacy. I remember that there were some cases where these values could be undefined, but I'm pretty sure I fixed that a while ago. This needs to be double checked, but I replaced all the `undefined` values by 0

File: lib/FragmentLoaderClassProvider.js [hunk 2]
```diff
@@ -100,17 +127,19 @@ function FragmentLoaderClassProvider(wrapper) {
...
 
-            const sendHttpRequestMetric = function(isSuccess, responseCode) {
+            const sendHttpRequestMetric = function(isSuccess, responseCode, requestDurationOffset = 0) {
 
-                request.firstByteDate = request.firstByteDate || request.requestStartDate;
-                request.requestEndDate = new Date();
+                let now = Date.now();
+                request.requestEndDate = new Date(now);
+                request.requestStartDate = new Date(now - requestDurationOffset);
+                request.firstByteDate = requestDurationOffset ? (new Date(now - requestDurationOffset + computeEstimatedRTT(requestDurationOffset))) : request.firstByteDate || request.requestStartDate;
```
Comment (AxelDelmas) [inline lib/FragmentLoaderClassProvider.js]: Not sure I understand the case where `requestDurationOffset === 0`:
From what I see in `onSuccess`, it equals 0 when there's no p2p download. Then `requestStartDate` should not be modified (and certainly not equal `requestEndDate` like here)

File: lib/FragmentLoaderClassProvider.js [hunk 3]
```diff
@@ -126,17 +155,20 @@ function FragmentLoaderClassProvider(wrapper) {
                     responseCode, // responseCode
                     request.duration, // mediaDuration
                     null, // responseHeaders // TODO: check if its really OK in all cases to pass null here
-                    isSuccess ? traces : null // traces
+                    isSuccess ? compensateTraceTimestamps(traces, requestDurationOffset) : null // traces
```
Comment (AxelDelmas) [inline lib/FragmentLoaderClassProvider.js]: These traces are "snapshots" of the progress events? Do you know where there used in dash.js?

File: lib/PlayerInterface.js [hunk 4]
```diff
@@ -35,7 +36,48 @@ class PlayerInterface {
...
+         * On the other side of the trade-off stands: If we reset the virtual buffer too defensively,
+         * we might choose a sub-optimal quality.
+         *
+         * NOTE: Current BOLA implementation uses buffer-level in its steady state
+         *       but will use a previous request stats to determine an appropriate startup bitrate.
+         *       Therefore a correct stats feedback from loaders also matter in this case.
+         *
+         * See: http://arxiv.org/abs/1601.06748
+         *
+         */
```
Comment (AxelDelmas) [inline lib/PlayerInterface.js]: Good explanation :+1: !

File: lib/FragmentLoaderClassProvider.js (hunk 1 above)
Comment (tchakabam) [inline lib/FragmentLoaderClassProvider.js]: great. then we could simplify that code in hlsjs-wrapper too

File: lib/FragmentLoaderClassProvider.js (hunk 3 above)
Comment (tchakabam) [inline lib/FragmentLoaderClassProvider.js]: Yes, they are allowing to comprehend the course of a download after hand. One could establish a bitrate-curve or so, if one would want to base an algorithm on that.

I don't think any of the rules use that level of precision yet. Also it doesn't seem of any importance what the absolute timestamp of the events was. But lets better make sure its all fitting together I thought.

Comment (tchakabam) [inline lib/FragmentLoaderClassProvider.js]: @AxelDelmas actually they do! that's what is being used to calculate throughput in all rules:

https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/rules/abr/ThroughputRule.js#L121
https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/rules/abr/BolaRule.js#L216

it gets converted from the metrics model trace to the `trace` property of the request itself here:

https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/models/MetricsModel.js#L140

File: lib/FragmentLoaderClassProvider.js (hunk 2 above)
Comment (tchakabam) [inline lib/FragmentLoaderClassProvider.js]: this should be fixed @AxelDelmas

**Comment by ruslandinov** [comment]
Except Axel's comment seems allright to me. 👍

**Pull request merged**